server/api/src/rsDataObjRepl.cpp

```diff
@@ -477,62 +481,88 @@ namespace
+                [[fallthrough]];
+            case static_cast<int>(access_type::delete_object): // NOLINT(bugprone-branch-clone)
+                [[fallthrough]];
+            case static_cast<int>(access_type::own):
+                return true;
+
+            default:
+                return false;
+        }
+    } // permission_is_sufficient_to_replicate_object
+
     auto user_has_permission_to_replicate_object(RsComm& _comm, const irods::file_object_ptr _obj) -> bool
     {
+        namespace adm = irods::experimental::administration;
+
         // Short-circuit the query by checking to see whether this is a rodsadmin using the ADMIN_KW, which allows
         // them to do whatever they want.
         if (irods::is_privileged_client(_comm) && getValByKey(&_obj->cond_input(), ADMIN_KW)) {
             return true;
         }
 
+        const auto user = adm::user{_comm.clientUser.userName, _comm.clientUser.rodsZone};
         const auto path = irods::experimental::filesystem::path{_obj->logical_path()};
-        const auto permission_query_string =
-            fmt::format("select DATA_ACCESS_TYPE where USER_NAME = '{}' and USER_ZONE = '{}' and DATA_ID = '{}'",
-                        _comm.clientUser.userName,
-                        _comm.clientUser.rodsZone,
-                        _obj->data_id());
-
-        auto permission_query = irods::query{&_comm, permission_query_string};
-
-        // If the user has no permissions on the object, it may have returned an empty result set.
-        if (permission_query.empty()) {
-            return false;
-        }
-
-        const auto& query_result = permission_query.front();
-        const auto& access_type_string = query_result[0];
 
         try {
-            using access_type = irods::experimental::catalog::access_type;
-
-            switch (std::stoi(access_type_string)) {
-                case static_cast<int>(access_type::modify_object): // NOLINT(bugprone-branch-clone)
-                    [[fallthrough]];
-                case static_cast<int>(access_type::delete_object): // NOLINT(bugprone-branch-clone)
-                    [[fallthrough]];
-                case static_cast<int>(access_type::own):
-                    return true;
-
-                default:
-                    return false;
+            // Check access permission for the user first...
+            const auto user_permission_query_string =
+                fmt::format("select DATA_ACCESS_TYPE where USER_NAME = '{}' and USER_ZONE = '{}' and DATA_ID = '{}'",
+                            user.name,
+                            user.zone,
+                            _obj->data_id());
+            auto user_permission_query = irods::query{&_comm, user_permission_query_string};
+            if (!user_permission_query.empty() &&
+                permission_is_sufficient_to_replicate_object(std::stoi(user_permission_query.front()[0])))
+            {
+                return true;
             }
```

Comment:
Is there a reason the permission check couldn't be done by the database? I think you can add an additional condition (i.e `DATA_ACCESS_TYPE = '<int_value>'`) and avoid the use of `std::stoi` altogether.

Not required, just pointing it out because it reduces the number of operations and removes an exception case.

Reply:
That's a good point. In that case, we also wouldn't need to loop through results for the group permissions because any result would indicate sufficient permissions.

I'm also now noticing that `permission_is_sufficient_to_replicate_object` does not account for `create_token`, `delete_token`, or `curate` permissions. However, `ichmod` does not advertise these permission levels and we don't seem to have tests for them, so I guess it's fine?

Anyway, with that in mind, do you have a preference between these two?

1. Leveraging the linear permission model:
```
DATA_ACCESS_TYPE >= '1120'
```

2. Checking for specific permission types:
```
DATA_ACCESS_TYPE = '1120' || = '1130' || = '1140' || = '1150' || = '1160' || = '1200'
```

I prefer 1, I think, because we assume the linear permission model in many places.

Reply:
I went ahead and implemented 1. We can iterate on that.